### PR TITLE
Refactor code base and build pipeline

### LINKED ISSUE #5: Return hash of published entry from `create()`, `update()` and `delete()` methods on `session`
Currently each of these methods return the session instance itself (`this`) but for when you need the id of the created entry, returning the entry hash would be very useful. I can see the reason for returning `this` for the setter methods (`setKeyPair()`, `setSchema()`) as it's nice to chain these, but I can't really see the case where we need it for create, update and delete.

### LINKED ISSUE #8: Make sure building pipeline accounts for all usage scenarios

### LINKED ISSUE #9: Revisit API

### PR DESCRIPTION
**Pull request by adzialocha**
Many things have changed since shirokuma got touched last: The `p2panda-js` API matured and our ideas around p2panda as well. Before we release the first version of shirokuma we want to bring it up to date with everything. This mostly concerns the "inners" of shirokuma, the actual, public API didn't change.

**This PR covers the following refactors:**

* Make use of native [private class fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) fields (!) in JavaScript ala `this.#keyPair` :exploding_head: 
* Remove (semi-automated) test fixtures and replace with hard-coded values. Maintaining this was a headache and writing simple test for shirokuma became unnecessarily complex
* Decouple concerns: Session API, GraphQL requests, entry signing etc. are all separated now
* Remove caching layer: Since we introduced querying the next entry arguments based on document view ids the current caching logic doesn't work anymore. Caching keys are currently organized as `${publicKey}/${viewId}`, which means that it will just make an entry in the cache after every next entry request but never hit afterwards as the view id changes constantly. To bring back caching we need to extend the API with something which allows us to handle document instances, this can be handled in future PRs: https://github.com/p2panda/shirokuma/issues/13
* Return "local document view id" instead of `Session` instance when creating, updating, deleting documents: https://github.com/p2panda/shirokuma/issues/5
* Removing "marshalling" logic, all of this is now handled by `p2panda-js`
* `p2panda-js` offers a "slim" and "inlined" version, where the latter has the WebAssembly code inlined. We want to also make use of these options in shirokuma. To enable this the rollup build pipeline has been improved
* Removed logging (for now): With the codebase becoming simpler now I couldn't find a reason to log every request and potential error, both of them you can see in your regular inspector tools

**Thoughts on the build pipeline:**

Currently we do _not_ bundle external dependencies in our CommonJS, ESM and NodeJS builds. _Only_ the UMD builds contain the whole thing. I couldn't find any "best practices" yet on how to do this but hopefully with a modern JS environment these decisions make sense: Webpack, Rollup etc. handles this automatically for us and if we want to go fancy ESM in the browser via `<script type="module">` we load the external dependencies manually before (will write something about it in the README.md). The UMD build should cover the needs of developers who "don't care" and simply just pull something in to play with.

**Todo**

- [x] Fix JSON-import assertion error of tsc
- [x] Find out why entries are differently encoded in js world .. (in the fixtures the log id was still starting at 1!)
- [x] Remove caching layer for now
- [x] Add back all the tests (w. refactored fixtures)
- [x] Test everything w. latest (unreleased) `p2panda-js` version
- [x] Add back all the doc strings

**Follow ups**

- https://github.com/p2panda/shirokuma/issues/6
- https://github.com/p2panda/shirokuma/issues/15
- https://github.com/p2panda/shirokuma/issues/14

Closes #8, #9, #5

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header

### CODE REVIEW

**Comment by adzialocha** [comment]
I'd like to write more about the different module flavours in the README.md and provide a nice modern browser example, but this can be handled in a separate PR

**Comment by adzialocha** [comment]
Coverage report is sad, but I'm not sure if winning the coverage award helps here .. all the other methods are barely wrappers around `p2panda-js` (which is tested). I guess adding some simple unit tests checking against the missing arguments will be enough. What do you think?

**Comment by sandreae** [comment]
> Coverage report is sad, but I'm not sure if winning the coverage award helps here .. all the other methods are barely wrappers around `p2panda-js` (which is tested). I guess adding some simple unit tests checking against the missing aruments will be enough. What do you think?

Yeh, I don't think this is necessarily the place to fight the coverage wars 😉

I can do a review of this PR tomorrow 👍 I'm looking forward to it!

File: README.md [hunk 1]
```diff
@@ -75,40 +72,28 @@ const keyPair = new KeyPair();
...
+or UMD module. It can easily be integrated into Webpack, Rollup or other tools.
+
+Since `shirokuma` contains WebAssembly code, it is necessary to initialise it
+before using the methods in the Browser. This initialisation step is not
+required in NodeJS contexts.
+
+To make this step a little bit easier `shirokuma` inlines the WebAssembly code
+as a base64 string which gets decoded automatically during initialisation. For
+manual initialisation the package also comes with "slim" versions where you
+need to provide a path to the ".wasm" file yourself.
```
Comment (sandreae) [inline README.md:95]: Wow, so much support :star_struck:

File: src/operation.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import type {
+  CreateArgs,
+  DeleteArgs,
+  EntryArgs,
+  PublishVariables,
+  UpdateArgs,
+} from './types';
+
+/**
+ * Signs and publishes a CREATE operation with the given data fields.
```
Comment (sandreae) [inline src/operation.ts]: This doesn't seem to publish the operation, only create it, is the doc string wrong?

Comment (sandreae) [reply src/operation.ts]: Same for other methods in this file.

File: src/session.ts [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+   * @returns Session instance
+   */
+  setKeyPair(keyPair: KeyPair): Session {
+    this.#keyPair = keyPair;
+    return this;
+  }
+
+  /**
+   * Return arguments for constructing the next entry given public key and
+   * schema id.
```
Comment (sandreae) [inline src/session.ts]: Should say "public key and document view id"

File: src/session.ts [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+  /**
+   * Publish an encoded entry and operation.
+   *
+   * This method returns the "local" document view id, which represents the
+   * latest version of the document we've created, updated or deleted.
+   *
+   * It is "local" because it is the "latest version" from our perspective.
+   * Concurrent updates through other peers might have happend but we didn't
+   * know about them in the moment we've published our operation. p2panda
+   * handles these concurrent updates internally for us.
```
Comment (sandreae) [inline src/session.ts]: :+1:

Comment (sandreae) [reply src/session.ts]: Maybe worth mentioning here that the current view id including any concurrent edits from other peers can be fetched through the query api.

File: src/session.ts [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+};
+
+/**
+ * Communicate with the p2panda network through a `Session` instance.
+ *
+ * `Session` provides a high-level interface to create data in the p2panda
+ * network by creating, updating and deleting documents following data schemas.
+ * It also provides a low-level API for directly accessing and creating
+ * entries on the Bamboo append-only log structure.
+ *
```
Comment (sandreae) [inline src/session.ts]: What does "accessing" mean in this case? Is that retrieving next args or is the meaning elsewhere?

**Review by sandreae — CHANGES_REQUESTED** [review]
This all looks really great, I like how the api is now exposing some of the lower-level functionality while still keeping the high level `Session` as the main interface.

The build pipeline is wild, fantastic to support so many targets and have a well considered flow for that.

My only comments were about clarification in doc strings I think.

File: src/session.ts (hunk 4 above)
Comment (adzialocha) [reply src/session.ts]: Can't think of a good way to describe this currently without writing too much. I think its a sign of that the current API does not really account for that flow "intuitively".

File: src/session.ts (hunk 5 above)
Comment (adzialocha) [reply src/session.ts]: Good question, I haven't written that sentence :-D But I understand it as "accessing" the low level API to create entries?

Comment (adzialocha) [reply src/session.ts]: I remove it as it's somehow implied by "... provides a low-level API"

File: README.md (hunk 1 above)
Comment (adzialocha) [reply README.md:95]: It's the same as we have in `p2panda-js` :partying_face:

File: src/session.ts (hunk 4 above)
Comment (sandreae) [reply src/session.ts]: How about something like: "If concurrent edits by other peers should also be referred to (as is often the case) then the document view id should be accessed via the document query api."

Comment (sandreae) [reply src/session.ts]: and remove the currently final sentence "p2panda handles these concurrent updates internally for us"

Comment (sandreae) [reply src/session.ts]: > I think its a sign of that the current API does not really account for that flow "intuitively".

Agree with this though.

Comment (adzialocha) [reply src/session.ts]: > How about something like: "If concurrent edits by other peers should also be referred to (as is often the case) then the document view id should be accessed via the document query api."

Nice, I like that! Thanks!

**Review by sandreae — APPROVED** [review]

**Pull request merged**